### PR TITLE
Propagate Context

### PR DESCRIPTION
**Pull request by leonnicolas**
Propagate a context from the main function to wherever it is needed.

Use `github.com/oklog/run` run groups to handle the life cycle of the go
routines running the rebooter and metrics server.

Ref: https://github.com/kubereboot/kured/issues/234
and https://github.com/kubereboot/kured/pull/808

Signed-off-by: leonnicolas <leonloechner@gmx.de>

### CODE REVIEW

**Review by evrardjp — COMMENTED** [review]
I think it's a good step. It implements what we talked about in the fast, with a real explanation of why we use context.

Yet, I am not the best person to judge wheter this is _enough_ or whether it's the right context. (And I sadly don't have enough time to delve into it).

Would you mind explaining why it mattered for you? And why did you want to use background context, please? And why you used oklog?

Thank you.

**Comment by leonnicolas** [comment]
> Would you mind explaining why it mattered for you? 

Passing down a context makes sure that when a caller wants to cancel an operation, some function down the line will not block until it returns. Instead it would probably return a context canceled. 
In practice, when kured received a SIGINT, it can cancel all outgoing requests and finish processing all incoming requests to the metrics server before it exits.

> And why did you want to use background context, please?

This is normally the context you use in the "root" of the application. https://pkg.go.dev/context#Background

> And why you used oklog?

At the moment kured starts 2 concurrent processes (go threads) in its main function and the metrics server that blocks the main thread forever.
If any of the go threads crash, we would continue to run the metrics server. No way to tell the other thread to exit as well and terminate the main thread (or try to restart the crashed thread). 
oklog's run package just provides one way to manage the life cycle of threads. We could probably use other packages as well. It was just my personal favorite because its API is so slim.

File: cmd/kured/main.go [hunk 1]
```diff
@@ -421,22 +425,23 @@ func (kb KubernetesBlockingChecker) isBlocked() bool {
 	return false
 }
 
-func rebootBlocked(blockers ...RebootBlocker) bool {
+func rebootBlocked(ctx context.Context, blockers ...RebootBlocker) bool {
+	// TODO: can we do that concurrently
```
Comment (jackfrancis) [inline cmd/kured/main.go:429]: So that we don't commit a `// TODO` comment, could you add an issue describing this optimization instead? (agree that this is out of scope for this PR)

thank you!
